### DISCUSSION TITLE
Classic worker with ESM app

### DISCUSSION THREAD

**Discussion by steverep** [webpack/webpack#17298] [Q&A] [answered]
I want to output ES modules for an app which uses a few workers.  However, I do not want the workers to use `import()` or `type: "module"`, i.e. they should remain "classic".  Reasons are:
1. Firefox support for ESM workers is still behind  a config flag, and until version 113 dynamic import was not supported
2. Experiencing hanging promises using `comlink` (which otherwise work fine with no ESM output)

The only way I see to attempt this is using `workerChunkLoading` as follows:

```js
const config = {
  target: "browserslist:modern", // supports dynamic import
  experiments: { outputModule: true },
  output: {
    module: true,
    workerChunkLoading: "import-scripts",
  }
}
```

However, this fails with a syntax error because the chunk format is still ESM for everything, even chunks only loaded by the worker.  Is there some way to force those chunks back to "array-push"?

It seems to me that a much more useful option would be `workerChunkFormat` and then set `workerChunkLoading` based on that, just as `chunkFormat` and `chunkLoading` behave.

**Comment by alexander-akait** [comment]
You can try `plugins: [new webpack.webworker.WebWorkerTemplatePlugin()]`, but I don't think it is what you want.

But I have some questions:
- Why you want to use classic script inside ESM?
- Do you have `import ... from ...` inside workers?

Because if you have we can't load ESM inside classic Worker, otherwise we need duplicate ESM and non ESM chunks, I don't think you want to do it.

The `workerChunkFormat` and `chunkFormat` may not be compatible, so as you can see there is no such option.

Alternative solution - multi compiler module, compile worker separately and them use it inside the main build

**Reply by steverep** [reply]
> You can try `plugins: [new webpack.webworker.WebWorkerTemplatePlugin()]`, but I don't think it is what you want.

It's unclear to me how to use this properly so I'd have to experiment.  Would I just specify `options.chunkFormat = "array-push"`?  Do I need to repeat all other output options?

> But I have some questions:
> 
>     * Why you want to use classic script inside ESM?

As stated in my OP, because Firefox doesn't support ESM workers without turning on a config flag, and `comlink` promises seem to just hang.


>     * Do you have `import ... from ...` inside workers?

Yes, but mainly just for `comlink`.  They are relatively tiny compared to the whole app.



> Because if you have we can't load ESM inside classic Worker, otherwise we need duplicate ESM and non ESM chunks, I don't think you want to do it.


That's exactly what I'd want.  The penalty for duplicating a small vendor package would be far over-shadowed by the speed gained by using ESM and native `import()` for the rest of the app.

> Alternative solution - multi compiler module, compile worker separately and them use it inside the main build

I considered that, but it's an awful lot of build script changes for something I would hope Webpack could handle.

**Reply by alexander-akait** [reply]
> It's unclear to me how to use this properly so I'd have to experiment. Would I just specify options.chunkFormat = "array-push"? Do I need to repeat all other output options?

Just add this to your configuration and you don't need to setup nothing extra, but code will not be in ESM format, that is why I think you don't want it

> That's exactly what I'd want. The penalty for duplicating a small vendor package would be far over-shadowed by the speed gained by using ESM and native import() for the rest of the app.

I am afraid webpack can't do it, because it was originally thought that such behavior is undesirable. Even adding `workerChunkFormat` will not solve it...

> I considered that, but it's an awful lot of build script changes for something I would hope Webpack could handle.

You don't need extra scripts, just set [`dependencies`](https://webpack.js.org/configuration/other-options/#dependencies) and webpack will run it before the main build.

Alternative solution - using babel/swc/etc code inliner, also you can inject code using `DefinePlugin`/`BannerPlugin` (they are hacks but will work and don't think we will break it, just non standard usage).

**Reply by steverep** [reply]
> Just add this to your configuration and you don't need to setup nothing extra, but code will not be in ESM format, that is why I think you don't want it


Without any options object, it does nothing to the workers and instead causes all `inport()` loading code to be replaced with `importScripts` i.e. for the entry chunks which of course doesn't make sense).  Specifying `module: false` or `chunkFormat: "array-push"` in the plugin options does nothing.

> I am afraid webpack can't do it, because it was originally thought that such behavior is undesirable. Even adding `workerChunkFormat` will not solve it...

😞 

> You don't need extra scripts, just set [`dependencies`](https://webpack.js.org/configuration/other-options/#dependencies) and webpack will run it before the main build.

Just so I know we're on the same page, you're talking about telling Webpack to ignore `new Worker()` (not sure how?), and then replacing with the output worker from the dependent config?

**Reply by alexander-akait** [reply]
> Without any options object, it does nothing to the workers and instead causes all inport() loading code to be replaced with importScripts i.e. for the entry chunks which of course doesn't make sense). Specifying module: false or chunkFormat: "array-push" in the plugin options does nothing.

Pluging don't have options, and override some of them to use `importScript`, but yes, as I said above it was not design for such usages, so I think multi compiler usage is the only one solution without hacks.

Not, just setup:
```
module.exports = [
  {
    name: 'my-worker',
    target: 'webworker',
    output: { path: path.resolve(__dirname, "worker-dir"), filename: "worker-[name].js"  }
    // …
  },
  {
    name: 'app',
    target: 'web',
    dependencies: ['my-worker'],
  },
];
```

and then in code use `new Worker(new URL("../worker-dir/worker-main.js", import.meta.url))`

Also there is a theoretical solution with `layer`, but it will require to use an extra entry, not `new Worker(...)` syntax and more options - https://webpack.js.org/configuration/entry-context/#entry-descriptor

**Reply by steverep** [reply]
> and then in code use `new Worker(new URL("../worker-dir/worker-main.js", import.meta.url))`

Won't the app config just then just re-bundle the worker and replace the URL?

> Also there is a theoretical solution with `layer`, but it will require to use an extra entry, not `new Worker(...)` syntax and more options - https://webpack.js.org/configuration/entry-context/#entry-descriptor

Yeah, I considered that as it looks like the only way to specify alternative chunk format in the same build but seems like too much of a pain.

The only other hack I was thinking might work would be to add a `splitChunks` rule to make sure the workers never have any chunks to load, and then go back and strip `{ type: "module" }` from the `new Worker()` call.

**Reply by alexander-akait** [reply, marked as answer]
> Won't the app config just then just re-bundle the worker and replace the URL?

Will but you can setup `noParse` to avoid it, but it is not a problem, somtimes you can have different babel/swc configurations for libraries/applications, so you can setup it based on what you want

> The only other hack I was thinking might work would be to add a splitChunks rule to make sure the workers never have any chunks to load, and then go back and strip { type: "module" } from the new Worker() call.

It might work, but that's right, it's a hack, so be careful when using it

I still recommend using multicompiler mode, it is not hard and more control and you don't need to modify commands and scripts, webpack will run everything under the hood

**Comment by steverep** [comment]
Trying the hack first to force workers not to have any imports, but I can't seem to affect it with `splitChunks`...

```js
      splitChunks: {
        cacheGroups: {
          workers: {
            test: /_worker/,  // should select worker files
            // chunks: () => false,  // no affect with or without this
            minChunks: 1000,  // just an absurdly high number
            reuseExistingChunk: false,
            idHint: "workers",
          },
        },
      },
```

That behaves like it doesn't even exist.  What am I doing wrong here?

**Reply by alexander-akait** [reply]
Not sure, but maybe, need to debug the split plugin to undestand it, try to set `reuseExistingChunk`, but it should work, also please use `chunks` to selector only the worker chunk (using function)

**Reply by steverep** [reply]
Tried your suggestions and like a hundred other variants including testing directly for the modules being split out.  Nothing had any effect on the web workers. 😕

**Reply by alexander-akait** [reply]
@steverep Maybe you can provide a simple case (github repo) and I will help you?

**Reply by steverep** [reply]
Thanks but I figured out a much simpler way to do it:

```js
      splitChunks: {
        // Disable splitting for web workers
        chunks: (chunk) =>
          !chunk.canBeInitial() && // same as "async" (default)
          !/(sort_filter|markdown)_worker/.test(chunk.name),
      },
```

Not sure why I didn't try that first.  Still don't understand why the cache group didn't work, but this is better. 🤷🏻‍♂️ 

On another note, I'll be filing an issue soon because it appears to me that web workers simply do not work with ESM output if they have any external chunks to load.  What I thought originally was `comlink` occurs with any external chunk - the runtime never gets past the `__webpack_require__` of those chunks.

**Reply by alexander-akait** [reply]
> On another note, I'll be filing an issue soon because it appears to me that web workers simply do not work with ESM output if they have any external chunks to load. What I thought originally was comlink occurs with any external chunk - the runtime never gets past the __webpack_require__ of those chunks.

For `__webpack_require__` we fixed a bug recently, it doesn't generate valid ESM output if you have outputModule, please update to the latest version

**Reply by steverep** [reply]
So I just tested 5.86 with the bug fix, and workers still have the same issue.

I was scanning the code and noticed that with `importScripts` loading, the external chunks are included as a startup dependency, and there's a bunch of extra runtime code.  Looks like ESM loading is the only method that doesn't include the startup dependency plugin.  Probably not the problem at all, but figured I'd mention it in case it were just that simple.

**Reply by alexander-akait** [reply]
Technically you can use `importScripts` inside your code and load something manually, if code is really small you can use `copy-webpack-plugin` and just copy this in your disk

**Reply by alexander-akait** [reply]
It will be minimized, so basic optimizations will be done

**Reply by steverep** [reply]
> Technically you can use `importScripts` inside your code and load something manually, 

The external chunks are a result of webpack code splitting, so there's no way to know them beforehand.  Disabling the splits is a fine workaround for now as our workers are pretty small.

So is the startup dependency plugin purposefully left out of ESM loading?

**Reply by alexander-akait** [reply]
> So is the startup dependency plugin purposefully left out of ESM loading?

I think it is a bug, there is an issue about it with `high` priority, hope we will solve it soon

**Comment by steverep** [comment]
Which exact issue are you referencing so I can follow it?  I see there are a number of them regarding workers.

**Reply by alexander-akait** [reply]
https://github.com/webpack/webpack/issues/17014
